### PR TITLE
[Requirement] Move validate php version check to CodeIgniter class

### PR DESCRIPTION
**Pull request by samsonasik**
By centralize check, no need to change in separate files between spark and public/index.php anymore.

**Checklist:**
- [x] Securely signed commits

### CODE REVIEW

**Review by MGatner — CHANGES_REQUESTED** [review]
1. This moves the check later in both entry points. Is there anything that would fail on lower PHP versions in between?
2. What does a dedicated class have to offer? We already track the framework version on `CodeIgniter\CodeIgniter`, seems like it might be a logical place to group this: `const CI_VERSION = '4.1.1';`

**Comment by samsonasik** [comment]
1. lower php version will be fail, eg in php 5.2, probably just use function and check very early (need require_once)
2. yes, that can be in CodeIgniter class, what do you prefer?

**Comment by MGatner** [comment]
I guess I'm not too worries about PHP < 7, as long as executed code will work up until the check in 7+. This will become more relevant as we drop versions moving forward, so for example if someone does not read the release notes and makes a point upgrade that breaks their app at least they will know why.

I would be in favor of consolidating this to `CodeIgniter` because of our framework version tracking there, but it's not a strong opinion. We can see what others think.

**Comment by samsonasik** [comment]
@MGatner I've moved the check to `CodeIgniter` class https://github.com/codeigniter4/CodeIgniter4/pull/4426/commits/077ca27ba12f72222a09d0da7578b2333c42e9ca

**Comment by samsonasik** [comment]
All green 🎉

File: system/Config/BaseService.php [hunk 1]
```diff
@@ -11,6 +11,20 @@
...
+use Config\Filters as ConfigFilters;
+use Config\Format as ConfigFormat;
+use Config\Honeypot as ConfigHoneyPot;
+use Config\Images;
+use Config\Migrations;
+use Config\Pager as ConfigPager;
+use Config\View as ConfigView;
+use CodeIgniter\HTTP\Request;
+use Config\Toolbar as ConfigToolbar;
+use Config\Validation as ConfigValidation;
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:27]: sort use statement

```suggestion
use Config\App;
use Config\Cache as CacheConfig;
use Config\Encryption as EncryptionConfig;
use Config\Exceptions as ExceptionsConfig;
use Config\Filters as FiltersConfig;
use Config\Format as FormatConfig;
use Config\Honeypot as HoneypotConfig;
use Config\Images as ImagesConfig;
use Config\Migrations as MigrationsConfig;
use Config\Pager as PagerConfig;
use Config\Toolbar as ToolbarConfig;
use Config\Validation as ValidationConfig;
use Config\View as ViewConfig;
use CodeIgniter\HTTP\Request;
```

File: system/Config/BaseService.php [hunk 2]
```diff
@@ -72,38 +86,38 @@
...
- * @method static Filters filters(\Config\Filters $config = null, $getShared = true)
- * @method static Format format(\Config\Format $config = null, $getShared = true)
- * @method static Honeypot honeypot(\Config\Honeypot $config = null, $getShared = true)
- * @method static BaseHandler image($handler = null, \Config\Images $config = null, $getShared = true)
+ * @method static EncrypterInterface encrypter(Encryption $config = null, $getShared = false)
+ * @method static Exceptions exceptions(ConfigExceptions $config = null, IncomingRequest $request = null, Response $response = null, $getShared = true)
+ * @method static Filters filters(ConfigFilters $config = null, $getShared = true)
+ * @method static Format format(ConfigFormat $config = null, $getShared = true)
+ * @method static Honeypot honeypot(ConfigHoneyPot $config = null, $getShared = true)
+ * @method static BaseHandler image($handler = null, Images $config = null, $getShared = true)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:100]:
```suggestion
 * @method static Exceptions exceptions(ExceptionsConfig $config = null, IncomingRequest $request = null, Response $response = null, $getShared = true)
 * @method static Filters filters(FiltersConfig $config = null, $getShared = true)
 * @method static Format format(FormatConfig $config = null, $getShared = true)
 * @method static Honeypot honeypot(HoneypotConfig $config = null, $getShared = true)
 * @method static BaseHandler image($handler = null, ImagesConfig $config = null, $getShared = true)
```

File: system/Config/BaseService.php [hunk 3]
```diff
@@ -72,38 +86,38 @@
...
- * @method static CURLRequest curlrequest($options = [], \CodeIgniter\HTTP\ResponseInterface $response = null, \Config\App $config = null, $getShared = true)
+ * @method static CURLRequest curlrequest($options = [], ResponseInterface $response = null, App $config = null, $getShared = true)
  * @method static Email email($config = null, $getShared = true)
- * @method static EncrypterInterface encrypter(\Config\Encryption $config = null, $getShared = false)
- * @method static Exceptions exceptions(\Config\Exceptions $config = null, \CodeIgniter\HTTP\IncomingRequest $request = null, \CodeIgniter\HTTP\Response $response = null, $getShared = true)
- * @method static Filters filters(\Config\Filters $config = null, $getShared = true)
- * @method static Format format(\Config\Format $config = null, $getShared = true)
- * @method static Honeypot honeypot(\Config\Honeypot $config = null, $getShared = true)
- * @method static BaseHandler image($handler = null, \Config\Images $config = null, $getShared = true)
+ * @method static EncrypterInterface encrypter(Encryption $config = null, $getShared = false)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:95]:
```suggestion
 * @method static EncrypterInterface encrypter(EncryptionConfig $config = null, $getShared = false)
```

File: system/Config/BaseService.php [hunk 4]
```diff
@@ -72,38 +86,38 @@
  * @see http://blog.ircmaxell.com/2015/11/simple-easy-risk-and-change.html
  * @see http://www.infoq.com/presentations/Simple-Made-Easy
  *
- * @method static CacheInterface cache(\Config\Cache $config = null, $getShared = true)
- * @method static CLIRequest clirequest(\Config\App $config = null, $getShared = true)
- * @method static CodeIgniter codeigniter(\Config\App $config = null, $getShared = true)
+ * @method static CacheInterface cache(Cache $config = null, $getShared = true)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:89]:
```suggestion
 * @method static CacheInterface cache(CacheConfig $config = null, $getShared = true)
```

File: system/Config/BaseService.php [hunk 5]
```diff
@@ -72,38 +86,38 @@
...
  * @method static Logger logger($getShared = true)
- * @method static MigrationRunner migrations(\Config\Migrations $config = null, \CodeIgniter\Database\ConnectionInterface $db = null, $getShared = true)
- * @method static Negotiate negotiator(\CodeIgniter\HTTP\RequestInterface $request = null, $getShared = true)
- * @method static Pager pager(\Config\Pager $config = null, \CodeIgniter\View\RendererInterface $view = null, $getShared = true)
- * @method static Parser parser($viewPath = null, \Config\View $config = null, $getShared = true)
- * @method static View renderer($viewPath = null, \Config\View $config = null, $getShared = true)
- * @method static IncomingRequest request(\Config\App $config = null, $getShared = true)
- * @method static Response response(\Config\App $config = null, $getShared = true)
- * @method static RedirectResponse redirectresponse(\Config\App $config = null, $getShared = true)
+ * @method static MigrationRunner migrations(Migrations $config = null, ConnectionInterface $db = null, $getShared = true)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:104]:
```suggestion
 * @method static MigrationRunner migrations(MigrationsConfig $config = null, ConnectionInterface $db = null, $getShared = true)
```

File: system/Config/BaseService.php [hunk 6]
```diff
@@ -72,38 +86,38 @@
...
+ * @method static Negotiate negotiator(RequestInterface $request = null, $getShared = true)
+ * @method static Pager pager(ConfigPager $config = null, RendererInterface $view = null, $getShared = true)
+ * @method static Parser parser($viewPath = null, ConfigView $config = null, $getShared = true)
+ * @method static View renderer($viewPath = null, ConfigView $config = null, $getShared = true)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:108]:
```suggestion
 * @method static Pager pager(PagerConfig $config = null, RendererInterface $view = null, $getShared = true)
 * @method static Parser parser($viewPath = null, ViewConfig $config = null, $getShared = true)
 * @method static View renderer($viewPath = null, ViewConfig $config = null, $getShared = true)
```

File: system/Config/BaseService.php [hunk 7]
```diff
@@ -72,38 +86,38 @@
...
- * @method static Router router(\CodeIgniter\Router\RouteCollectionInterface $routes = null, \CodeIgniter\HTTP\Request $request = null, $getShared = true)
- * @method static Security security(\Config\App $config = null, $getShared = true)
- * @method static Session session(\Config\App $config = null, $getShared = true)
+ * @method static Router router(RouteCollectionInterface $routes = null, Request $request = null, $getShared = true)
+ * @method static Security security(App $config = null, $getShared = true)
+ * @method static Session session(App $config = null, $getShared = true)
  * @method static Throttler throttler($getShared = true)
  * @method static Timer timer($getShared = true)
- * @method static Toolbar toolbar(\Config\Toolbar $config = null, $getShared = true)
+ * @method static Toolbar toolbar(ConfigToolbar $config = null, $getShared = true)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:118]:
```suggestion
 * @method static Toolbar toolbar(ToolbarConfig $config = null, $getShared = true)
```

File: system/Config/BaseService.php [hunk 8]
```diff
@@ -72,38 +86,38 @@
...
  * @method static URI uri($uri = null, $getShared = true)
- * @method static Validation validation(\Config\Validation $config = null, $getShared = true)
+ * @method static Validation validation(ConfigValidation $config = null, $getShared = true)
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:120]:
```suggestion
 * @method static Validation validation(ValidationConfig $config = null, $getShared = true)
```

File: system/Config/BaseService.php [hunk 9]
```diff
@@ -11,6 +11,20 @@
...
+use Config\Images;
+use Config\Migrations;
+use Config\Pager as ConfigPager;
+use Config\View as ConfigView;
+use CodeIgniter\HTTP\Request;
+use Config\Toolbar as ConfigToolbar;
+use Config\Validation as ConfigValidation;
 use CodeIgniter\Autoloader\Autoloader;
 use CodeIgniter\Autoloader\FileLocator;
 use CodeIgniter\CodeIgniter;
```
Comment (mostafakhudair) [inline system/Config/BaseService.php:30]:
```suggestion
use CodeIgniter\Autoloader\Autoloader;
use CodeIgniter\Autoloader\FileLocator;
use CodeIgniter\CodeIgniter;
use Config\App;
use Config\Cache as CacheConfig;
use Config\Encryption as EncryptionConfig;
use Config\Exceptions as ExceptionsConfig;
use Config\Filters as FiltersConfig;
use Config\Format as FormatConfig;
use Config\Honeypot as HoneypotConfig;
use Config\Images as ImagesConfig;
use Config\Migrations as MigrationsConfig;
use Config\Pager as PagerConfig;
use Config\Toolbar as ToolbarConfig;
use Config\Validation as ValidationConfig;
use Config\View as ViewConfig;
```

Comment (mostafakhudair) [reply system/Config/BaseService.php:30]: sorted statement

**Review by mostafakhudair — CHANGES_REQUESTED** [review]

**Review by MGatner — APPROVED** [review]
The version constraint looks good, thank you for moving it! I'm not sure what has happened with the explosion of use statement changes but I would prefer to see all that in a separate PR.

**Comment by mostafakhudair** [comment]
@MGatner CLI is accessible without php version check

File: system/CodeIgniter.php [hunk 10]
```diff
@@ -147,6 +152,17 @@ class CodeIgniter
 	 */
 	public function __construct(App $config)
 	{
+		if (version_compare(PHP_VERSION, self::MIN_PHP_VERSION, '<'))
+		{
+			die(
+				sprintf(
+					'Your PHP version must be %s or higher to run CodeIgniter. Current version: %s',
```
Comment (totoprayogo1916) [inline system/CodeIgniter.php:159]: Is it possible, if this is added to Lang?

**Comment by samsonasik** [comment]
@MGatner I create separate PR for rector notice https://github.com/codeigniter4/CodeIgniter4/pull/4430

File: system/CodeIgniter.php (hunk 10 above)
Comment (samsonasik) [reply system/CodeIgniter.php:159]: @totoprayogo1916 if you mean multi for multi language, I am not sure since it early executed.

**Review by paulbalandan — APPROVED** [review]

**Comment by samsonasik** [comment]
rebased.

File: system/CodeIgniter.php (hunk 10 above)
Comment (MGatner) [reply system/CodeIgniter.php:159]: `lang()` is definitely not available yet. In general I'm okay with strictly internal messages being in English. I understand there are some internationalization efforts for PHP core but I think this is still pretty common?

**Comment by MGatner** [comment]
Looks much better! Thanks for splitting that. What about @mostafakhudair's concern that CLI execution bypasses version check?

**Comment by samsonasik** [comment]
@mostafakhudair any step to reproduce for by pass via cli ? I can't reproduce via spark command.

**Comment by mostafakhudair** [comment]
@samsonasik @MGatner everything is working well now, I didn't know what was the problem.

**Pull request merged**